### PR TITLE
[Hub Generated] Review request for Microsoft.Devices to add version preview/2023-03-01-preview

### PR DESCRIPTION
**Pull request by WendyCaiMS**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/WendyCaiMS-deviceprovisioningservices-Microsoft.Devices-2023-03-01-preview...main/).
## ARM API Information (Control Plane)

Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
        04/28/2023
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
        04/28/2023
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by zzvswxy** [comment]
Hi @WendyCaiMS  as there are LintDiff errors, can you fix it?
And for ApiReadinessCheck error,  is service release date confirmed? Or we can merge the PR after the service release? Thanks

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2023-03-01-preview/iotdps.json
Comment (rkmanda) [inline specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2023-03-01-preview/iotdps.json:2529]: >ArmIdentity [](http://example.com/codeflow?start=5&length=11)

Theres a common type for managed identities. Please use that.

Comment (rkmanda) [reply specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2023-03-01-preview/iotdps.json:2529]: [https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/managedidentity.json](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/resource-management/v5/managedidentity.json)

Comment (WendyCaiMS) [reply specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2023-03-01-preview/iotdps.json:2529]: Do you mean that our ArmIdentity and ArmUserIdentity definitions need to follow the same format as the managedidentity.json sent in the link? Does the naming also need to be changed to "ManagedServiceIdentity" etc?

**Comment by WendyCaiMS** [comment]
> 



> Hi @WendyCaiMS as there are LintDiff errors, can you fix it? And for ApiReadinessCheck error, is service release date confirmed? Or we can merge the PR after the service release? Thanks

Hi @zzvswxy,
For the Lindiff error: PatchdentityProperty, we do not want to support having the identity in our PATCH call because it will not be used. Our PATCH call is only used for updating tags. Can we get an override on this?
For the ApiReadinessCheck error, can we proceed with the merge if we provide a release date or does the service need to be deployed before the PR can be merged?
Thanks!

**Comment by zzvswxy** [comment]
Hi @WendyCaiMS  
When the RP finish ARM review, I will ask SDK owner to review the SDK breaking change. 

> For the ApiReadinessCheck error, can we proceed with the merge if we provide a release date or does the service need to be deployed before the PR can be merged?

We suggest merge the PR when service are deployed. But you can provide a release date for merging this PR.

> For the Lindiff error: PatchdentityProperty, we do not want to support having the identity in our PATCH call because it will not be used. Our PATCH call is only used for updating tags. Can we get an override on this?

Hi @jianyexi  , can you help me to reply this question? Thanks

**Comment by jianyexi** [comment]
> Hi @zzvswxy, For the Lindiff error: PatchdentityProperty, we do not want to support having the identity in our PATCH call because it will not be used. Our PATCH call is only used for updating tags. Can we get an override on this? For the ApiReadinessCheck error, can we proceed with the merge if we provide a release date or does the service need to be deployed before the PR can be merged? Thanks!

@rkmanda could you give your perspective, as it actually violates the RPC-Patch-V1-11

**Comment by WendyCaiMS** [comment]
Hi @rkmanda, just want to follow up on the previous questions regarding getting an override for the PatchIdentityProperty lintdiff error, and guidelines on how to use the common managed identity type in our swagger you shared.

**Comment by pilor** [comment]
PATCH is required for managed identity to ensure consistency across Azure.  Its not optional.  It must also follow this behavior:  https://eng.ms/docs/products/arm/authorization/managed_identities/msionboardingsteps#6-update-your-control-plane-putpatch-method

**Comment by WendyCaiMS** [comment]
> PATCH is required for managed identity to ensure consistency across Azure. Its not optional. It must also follow this behavior: https://eng.ms/docs/products/arm/authorization/managed_identities/msionboardingsteps#6-update-your-control-plane-putpatch-method
> 
> Its equivalent to a resource not allowing PATCHing of tags. Users expect it to work and it must work

Hi @pilor, thanks for your feedback. I am getting a page not found when going to the link you provided. Can you resend?

**Comment by pilor** [comment]
> > PATCH is required for managed identity to ensure consistency across Azure. Its not optional. It must also follow this behavior: https://eng.ms/docs/products/arm/authorization/managed_identities/msionboardingsteps#6-update-your-control-plane-putpatch-method
> > Its equivalent to a resource not allowing PATCHing of tags. Users expect it to work and it must work
> 
> Hi @pilor, thanks for your feedback. I am getting a page not found when going to the link you provided. Can you resend?

https://eng.ms/docs/products/arm/rbac/managed_identities/msionboardingsteps#6-update-your-control-plane-putpatch-method

**Comment by WendyCaiMS** [comment]
> PATCH is required for managed identity to ensure consistency across Azure. Its not optional. It must also follow this behavior: https://eng.ms/docs/products/arm/authorization/managed_identities/msionboardingsteps#6-update-your-control-plane-putpatch-method
> 
> Its equivalent to a resource not allowing PATCHing of tags. Users expect it to work and it must work

Hi @pilor, I want to make sure my understanding is correct. Currently, our PATCH call only supports updating tags. Any other updates to the resource, including identity information, is handled by our PUT. In order to adhere to the PatchIdentityProperty check, we can add "identity" as a property in the request body of the PATCH, but it will not actually be used. Is that okay?

**Comment by pilor** [comment]
> > PATCH is required for managed identity to ensure consistency across Azure. Its not optional. It must also follow this behavior: https://eng.ms/docs/products/arm/authorization/managed_identities/msionboardingsteps#6-update-your-control-plane-putpatch-method
> > Its equivalent to a resource not allowing PATCHing of tags. Users expect it to work and it must work
> 
> Hi @pilor, I want to make sure my understanding is correct. Currently, our PATCH call only supports updating tags. Any other updates to the resource, including identity information, is handled by our PUT. In order to adhere to the PatchIdentityProperty check, we can add "identity" as a property in the request body of the PATCH, but it will not actually be used. Is that okay?

You actually need to support PATCH on the identity properties in your API.

**Comment by WendyCaiMS** [comment]
> > > PATCH is required for managed identity to ensure consistency across Azure. Its not optional. It must also follow this behavior: https://eng.ms/docs/products/arm/authorization/managed_identities/msionboardingsteps#6-update-your-control-plane-putpatch-method
> > > Its equivalent to a resource not allowing PATCHing of tags. Users expect it to work and it must work
> > 
> > 
> > Hi @pilor, I want to make sure my understanding is correct. Currently, our PATCH call only supports updating tags. Any other updates to the resource, including identity information, is handled by our PUT. In order to adhere to the PatchIdentityProperty check, we can add "identity" as a property in the request body of the PATCH, but it will not actually be used. Is that okay?
> 
> You actually need to support PATCH on the identity properties in your API.

Hi @pilor,
Is the Patch support a requirement for preview? Can we move ahead with this version if we have plans to include Patch support in the GA version?

**Comment by mentat9** [comment]
> > > > PATCH is required for managed identity to ensure consistency across Azure. Its not optional. It must also follow this behavior: https://eng.ms/docs/products/arm/authorization/managed_identities/msionboardingsteps#6-update-your-control-plane-putpatch-method
> > > > Its equivalent to a resource not allowing PATCHing of tags. Users expect it to work and it must work
> > > 
> > > 
> > > Hi @pilor, I want to make sure my understanding is correct. Currently, our PATCH call only supports updating tags. Any other updates to the resource, including identity information, is handled by our PUT. In order to adhere to the PatchIdentityProperty check, we can add "identity" as a property in the request body of the PATCH, but it will not actually be used. Is that okay?
> > 
> > 
> > You actually need to support PATCH on the identity properties in your API.
> 
> Hi @pilor, Is the Patch support a requirement for preview? Can we move ahead with this version if we have plans to include Patch support in the GA version?

Rules are the same for preview vs. GA: from ARM perspective they are the same.

**Comment by WendyCaiMS** [comment]
Hi @mentat9, just for clarification, according to step 6 in this onboarding document: https://armwiki.azurewebsites.net/authorization/managed_identities/MSIOnboardingSteps.html is Patch support only for adding/removing user assigned identities and not system assigned?

**Comment by WendyCaiMS** [comment]
We evaluated the changes needed to implement PATCH support for identity and it will result in breaking backwards compatibility for our current PATCH call. Our PATCH currently is a synchronous call but implementing support for identity will require us to change it to asynchronous. Our service is also closely tied to IoTHub's services and they do not have PATCH support for identity and we want to keep the customer experience consistent. Can we get an exception to the PATCH support?

**Comment by TimLovellSmith** [comment]
@WendyCaiMS you can open a breaking changes request and or contact Breaking Changes Reviewers to discuss changing sync to async.

**Comment by TimLovellSmith** [comment]
Based on the non-optionality mentioned by pilor et al, calling this 'ARMChangesRequested' still.

**Comment by anushprabhu** [comment]
> @WendyCaiMS you can open a breaking changes request and or contact Breaking Changes Reviewers to discuss changing sync to async.

@TimLovellSmith – We see some services are not supporting PATCH for this functionality. Do you know if this is a recent change in the requirements for MSI onboarding for RPs?

**Comment by TimLovellSmith** [comment]
>>@WendyCaiMS you can open a breaking changes request and or contact Breaking Changes Reviewers to discuss changing sync to async.

>@TimLovellSmith – We see some services are not supporting PATCH for this functionality. Do you know if this is a recent change in the requirements for MSI onboarding for RPs?

@anushprabhu No, I do not know. /cc @pilor

**Comment by WendyCaiMS** [comment]
Hi @zzvswxy, we got the ARM signoff and our service release date is targeted for 04/28/2023. What are next steps for the PR approval process?

**Comment by zzvswxy** [comment]
Hi @lirenhe  can you help me for following the PR. Thanks

**Comment by WendyCaiMS** [comment]
> Hi @lirenhe can you help me for following the PR. Thanks

Hi @zzvswxy, @lirenhe just wanted to follow up on this PR. Our target release date is 04/28/2023. What else do we need to get approval for this PR?

**Review by lirenhe — APPROVED** [review]

**Pull request merged**
